### PR TITLE
Users/ajya/azurestacktasksupport

### PR DESCRIPTION
**Pull request by Ajay-MS**

### CODE REVIEW

File: Tasks/Common/VstsAzureHelpers_/InitializeFunctions.ps1 [hunk 1]
```diff
@@ -224,3 +229,82 @@ function CmdletHasMember {
...
+
+    $AzureKeyVaultDnsSuffix="vault.$($stackdomain)".ToLowerInvariant()
+    $AzureKeyVaultServiceEndpointResourceId= $("https://vault.$stackdomain".ToLowerInvariant())
+    $StorageEndpointSuffix = ($stackdomain).ToLowerInvariant()
+
+    # Check if endpoint data contains required data.
+    if($Endpoint.data.GraphUrl -eq $null)
+    {
+        Write-Verbose "Retrieving endpoints from the $ResourceManagerEndpoint"
+        $endpointData = Invoke-RestMethod -Method Get -Uri "$($EndpointURI.ToString().TrimEnd('/'))/metadata/endpoints?api-version=2015-01-01" -ErrorAction Stop
```
Comment (kmkumaran) [inline Tasks/Common/VstsAzureHelpers_/InitializeFunctions.ps1]: Does it handle proxy?

File: Tasks/Common/azurestack-common/azurestackrestutility.ts [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+            if(obj.portalEndpoint) {
+                endpoint['portalEndpoint'] = obj.portalEndpoint;
+            }
+            
+            deferred.resolve(endpoint);
+        }
+        else {
+            tl.debug(body);
+            deferred.reject(response.statusMessage);
+        }
```
Comment (kmkumaran) [inline Tasks/Common/azurestack-common/azurestackrestutility.ts]: Aren't KeyValutDnsSuffix and Storagesuffix not required?

File: Tasks/AzureResourceGroupDeployment/models/DeployAzureRG.ts [hunk 3]
```diff
@@ -106,14 +76,64 @@ export class AzureRGTaskParameters {
...
+        } else {
+            envAuthorityUrl = (envAuthorityUrl != null) ? envAuthorityUrl : "https://login.windows.net/";
+            activeDirectoryResourceId = armUrl;
+        }
+
+        var credentials = new msRestAzure.ApplicationTokenCredentials(servicePrincipalId, tenantId, servicePrincipalKey, armUrl, envAuthorityUrl, activeDirectoryResourceId);
         return credentials;
     }
+
+    public getAzureRGTaskParameters() : Promise<AzureRGTaskParameters>
```
Comment (Ajay-MS) [inline Tasks/AzureResourceGroupDeployment/models/DeployAzureRG.ts]: I have requested deepak for review

File: Tasks/AzureResourceGroupDeployment/package.json [hunk 4]
```diff
@@ -1,10 +1,7 @@
-
 {
   "name": "AzureResourceGroupDeployment",
-  "main":"main.js",
+  "main": "main.js",
   "dependencies": {
-    "vsts-task-lib" : "^0.9.20",
```
Comment (Ajay-MS) [inline Tasks/AzureResourceGroupDeployment/package.json]: Because these modules will be imported from common module. If I mentioned same module in common as well as in task it doesn't work properly.

File: Tasks/Common/azurestack-common/azurestackrestutility.ts [hunk 5]
```diff
@@ -0,0 +1,63 @@
...
+            if(authenticationData) {
+                var loginEndpoint = authenticationData.loginEndpoint;
+                if(loginEndpoint) {
+                    endpoint['activeDirectoryAuthority'] = loginEndpoint;
+                    endpoint['environmentAuthorityUrl'] = loginEndpoint;
+                }
+
+                var audiences = authenticationData.audiences;
+                if(audiences) {
+                    if(audiences.lenth > 0) {
```
Comment (vincentdass) [inline Tasks/Common/azurestack-common/azurestackrestutility.ts]: should it be audiences.length ?

**Comment by vincentdass** [comment]
@Ajay-MS , I could see many package-lock.json in Tests folder of Task.
Is it actually required ?

File: Tasks/Common/azurestack-common/package.json [hunk 6]
```diff
@@ -0,0 +1,21 @@
...
+  "author": "Microsoft Corporation",
+  "license": "MIT",
+  "bugs": {
+    "url": "https://github.com/Microsoft/vsts-tasks/issues"
+  },
+  "homepage": "https://github.com/Microsoft/vsts-tasks#readme",
+  "dependencies": {
+    "q": "1.4.1",
+    "vso-node-api": "6.0.1-preview",
+    "vsts-task-lib": "^0.9.20",
```
Comment (vincentdass) [inline Tasks/Common/azurestack-common/package.json]: remove `^` , (to use specific task version).

File: Tasks/Common/azurestack-common/package.json [hunk 7]
```diff
@@ -0,0 +1,21 @@
+{
+  "name": "azurestack-common",
+  "version": "1.0.0",
+  "description": "Common Library for Azure Rest Calls",
```
Comment (vincentdass) [inline Tasks/Common/azurestack-common/package.json]: Common Library for Azure **Stack** Rest Calls ?

File: Tasks/Common/azurestack-common/azurestackrestutility.ts [hunk 8]
```diff
@@ -0,0 +1,63 @@
...
+            
+            if(obj.portalEndpoint) {
+                endpoint['portalEndpoint'] = obj.portalEndpoint;
+            }
+            
+            deferred.resolve(endpoint);
+        }
+        else {
+            tl.debug(body);
+            deferred.reject(response.statusMessage);
```
Comment (vincentdass) [inline Tasks/Common/azurestack-common/azurestackrestutility.ts]: Can you please add some error Message here ?

File: Tasks/AzureResourceGroupDeployment/main.ts [hunk 9]
```diff
@@ -6,25 +6,29 @@ import virtualMachine = require("./operations/VirtualMachine");
...
+            case "Start":
+            case "Stop":
+            case "Restart":
+            case "Delete":
+            case "StopWithDeallocate":
+                return virtualMachineOperation.execute();
+            default:
+                throw tl.loc("InvalidAction", taskParameters.action);
+        }
+    }).catch((error) => {
```
Comment (Ajay-MS) [inline Tasks/AzureResourceGroupDeployment/main.ts]: Directly we can't call a async method synchronously from a sync method. To achieve this all operations are being performed after promise has been resolved.

File: Tasks/Common/VstsAzureHelpers_/InitializeFunctions.ps1 (hunk 1 above)
Comment (Ajay-MS) [inline Tasks/Common/VstsAzureHelpers_/InitializeFunctions.ps1]: Yes

File: Tasks/Common/VstsAzureHelpers_/InitializeFunctions.ps1 [hunk 10]
```diff
@@ -224,3 +229,82 @@ function CmdletHasMember {
         return false;
     }
 }
+
+<#
+    Adds Azure Stack environment to use with AzureRM command-lets when targeting Azure Stack
+#>
+function Add-AzureStackAzureRmEnvironment {
+    param (
+        [Parameter(mandatory=$true, HelpMessage="The Admin ARM endpoint of the Azure Stack Environment")]
```
Comment (Ajay-MS) [inline Tasks/Common/VstsAzureHelpers_/InitializeFunctions.ps1]: This HelpMessage is only used as a purpose of get-help.

File: Tasks/Common/azure-arm-rest/azure-arm-common.ts [hunk 11]
```diff
@@ -15,9 +15,10 @@ export class ApplicationTokenCredentials {
     private secret: string;
     public armUrl: string;
     public authorityUrl: string;
+    public activeDirectoryResourceId: string;
     private token_deferred: Q.Promise<string>;
 
-    constructor(clientId: string, domain: string, secret: string, armUrl: string, authorityUrl: string) {
+    constructor(clientId: string, domain: string, secret: string, armUrl: string, authorityUrl: string, activeDirectoryResourceId: string) {
         if (!Boolean(clientId) || typeof clientId.valueOf() !== 'string') {
```
Comment (sachinma) [inline Tasks/Common/azure-arm-rest/azure-arm-common.ts]: armUrl is no longer used in this class. Consider removing it. or rename it appropriately instead of adding new field

File: Tasks/Common/azure-arm-rest/azure-arm-network.ts [hunk 12]
```diff
@@ -14,7 +14,7 @@ export class NetworkManagementClient extends azureServiceClient.ServiceClient {
     constructor(credentials: msRestAzure.ApplicationTokenCredentials, subscriptionId, baseUri?: any, options?: any) {
         super(credentials, subscriptionId);
 
-        this.apiVersion = '2016-09-01';
+        this.apiVersion = '2015-06-15';
```
Comment (sachinma) [inline Tasks/Common/azure-arm-rest/azure-arm-network.ts]: why this change?

File: Tasks/Common/azurestack-common/azurestackrestutility.ts [hunk 13]
```diff
@@ -0,0 +1,76 @@
+var parseString = require('xml2js').parseString;
+
+import tl = require('vsts-task-lib/task');
+import Q = require('q');
+import httpClient = require('vso-node-api/HttpClient');
+import restClient = require('vso-node-api/RestClient');
+var deasync = require('deasync');
```
Comment (sachinma) [inline Tasks/Common/azurestack-common/azurestackrestutility.ts]: Are you using it?

File: Tasks/Common/azurestack-common/azurestackrestutility.ts [hunk 14]
```diff
@@ -0,0 +1,76 @@
...
+            if(obj.portalEndpoint) {
+                endpoint['portalEndpoint'] = obj.portalEndpoint;
+            }
+            
+            var endpointUrl =  endpoint.url;
+            endpointUrl += (endpointUrl.lastIndexOf("/") == endpointUrl.length-1) ? "":"/";
+            var domain = "";
+            try {
+                var index = endpointUrl.indexOf('.');
+                domain = endpointUrl.substring(index+1);
```
Comment (sachinma) [inline Tasks/Common/azurestack-common/azurestackrestutility.ts]: consider if url.parse can be used here

File: Tasks/Common/azurestack-common/azurestackrestutility.ts [hunk 15]
```diff
@@ -0,0 +1,76 @@
...
+import tl = require('vsts-task-lib/task');
+import Q = require('q');
+import httpClient = require('vso-node-api/HttpClient');
+import restClient = require('vso-node-api/RestClient');
+var deasync = require('deasync');
+
+var httpObj = new httpClient.HttpCallbackClient(tl.getVariable("AZURE_HTTP_USER_AGENT"));
+var restObj = new restClient.RestCallbackClient(httpObj);
+
+export async function initializeAzureStackData(endpoint)
```
Comment (sachinma) [inline Tasks/Common/azurestack-common/azurestackrestutility.ts]: Try defining types if possible in the signature

File: Tasks/AzureResourceGroupDeployment/package.json (hunk 4 above)
Comment (sachinma) [inline Tasks/AzureResourceGroupDeployment/package.json]: Consider using the make.json file to remove the nested vsts-lib

File: Tasks/AzureResourceGroupDeployment/models/DeployAzureRG.ts [hunk 16]
```diff
@@ -106,14 +76,72 @@ export class AzureRGTaskParameters {
...
         var tenantId: string = tl.getEndpointAuthorizationParameter(connectedService, "tenantid", false);
         var armUrl: string = tl.getEndpointUrl(connectedService, true);
         var envAuthorityUrl: string = tl.getEndpointDataParameter(connectedService, 'environmentAuthorityUrl', true);
-        envAuthorityUrl = (envAuthorityUrl != null) ? envAuthorityUrl : "https://login.windows.net/";
-        var credentials = new msRestAzure.ApplicationTokenCredentials(servicePrincipalId, tenantId, servicePrincipalKey, armUrl, envAuthorityUrl);
+        var environment: string = tl.getEndpointDataParameter(connectedService, 'environment', true);
+        var activeDirectoryResourceId: string = tl.getEndpointDataParameter(connectedService, 'activeDirectoryServiceEndpointResourceId', true);
+        
+        if(environment != null && environment.toLowerCase() == azureStackEnvironment.toLowerCase()) {
+            if(!envAuthorityUrl || !activeDirectoryResourceId) {
```
Comment (sachinma) [inline Tasks/AzureResourceGroupDeployment/models/DeployAzureRG.ts]: what is activeDirectoryResourceId and what is armURL (in this class). tl.getEndpointUrl(connectedService, true) - not useful anymore?

File: Tasks/AzureResourceGroupDeployment/models/DeployAzureRG.ts [hunk 17]
```diff
@@ -106,14 +76,72 @@ export class AzureRGTaskParameters {
         }
     }
 
-    private getARMCredentials(connectedService: string): msRestAzure.ApplicationTokenCredentials {
+    private  async getARMCredentials(connectedService: string){
```
Comment (sachinma) [inline Tasks/AzureResourceGroupDeployment/models/DeployAzureRG.ts]: You should modify the signature to specify the return type i.e. Promise<msrestAzure...>

File: Tasks/Common/azure-arm-rest/azure-arm-network.ts (hunk 12 above)
Comment (Ajay-MS) [inline Tasks/Common/azure-arm-rest/azure-arm-network.ts]: Latest api version being supported by Azure Stack is 2015-06-01 that's why had to decrease API version.

File: Tasks/Common/azure-arm-rest/azure-arm-common.ts (hunk 11 above)
Comment (Ajay-MS) [inline Tasks/Common/azure-arm-rest/azure-arm-common.ts]: armUrl  will be still in use as base url for all REST request to azure - https://github.com/Microsoft/vsts-tasks/blob/master/Tasks/Common/azure-arm-rest/AzureServiceClient.ts#L77

File: Tasks/AzureResourceGroupDeployment/models/DeployAzureRG.ts (hunk 16 above)
Comment (Ajay-MS) [inline Tasks/AzureResourceGroupDeployment/models/DeployAzureRG.ts]: armUrl - It acts as base url to make any REST request to azure. 
activeDirectoryResourceId - Used to fetch access token as app in active directory get registered with this Url. In case of all national cloud armUrl and activeDirectoryResourceId urls are same. But in case of Azure Stack both url differs.

File: Tasks/Common/azurestack-common/azurestackrestutility.ts (hunk 14 above)
Comment (Ajay-MS) [inline Tasks/Common/azurestack-common/azurestackrestutility.ts]: Here we need suffix of base url,  url.parse will not help.

File: Tasks/Common/azure-arm-rest/azure-arm-network.ts (hunk 12 above)
Comment (sachinma) [inline Tasks/Common/azure-arm-rest/azure-arm-network.ts]: Check this one out. Talk to Deepak once he is in.

File: Tasks/Common/azure-arm-rest/azure-arm-common.ts [hunk 18]
```diff
@@ -38,13 +39,19 @@ export class ApplicationTokenCredentials {
             throw new Error(tl.loc("authorityUrlCannotBeEmpty"));
         }
 
+        if (!Boolean(activeDirectoryResourceId) || typeof activeDirectoryResourceId.valueOf() !== 'string') {
+            throw new Error(tl.loc("activeDirectoryResourceIdUrlCannotBeEmpty"));
```
Comment (sachinma) [inline Tasks/Common/azure-arm-rest/azure-arm-common.ts]: Did you add this loc string

File: Tasks/AzureResourceGroupDeployment/main.ts [hunk 19]
```diff
@@ -6,25 +6,29 @@ import virtualMachine = require("./operations/VirtualMachine");
...
+        throw error;
```
Comment (sachinma) [inline Tasks/AzureResourceGroupDeployment/main.ts]: You can leave the catch block; since it will be handled in caller

File: Tasks/AzureResourceGroupDeployment/models/DeployAzureRG.ts [hunk 20]
```diff
@@ -106,14 +76,72 @@ export class AzureRGTaskParameters {
...
+            activeDirectoryResourceId = armUrl;
+        }
+
+        var credentials = new msRestAzure.ApplicationTokenCredentials(servicePrincipalId, tenantId, servicePrincipalKey, armUrl, envAuthorityUrl, activeDirectoryResourceId);
         return credentials;
     }
+
+    public getAzureRGTaskParameters() : Promise<AzureRGTaskParameters> 
+    {
+        return new Promise<AzureRGTaskParameters>(async (resolve, reject) => {
```
Comment (sachinma) [inline Tasks/AzureResourceGroupDeployment/models/DeployAzureRG.ts]: P2: Consider making getAzureRGTaskParameters async and removing this new Promise

File: Tasks/Common/azure-arm-rest/azure-arm-resource.ts [hunk 21]
```diff
@@ -12,7 +12,7 @@ export class ResourceManagementClient extends azureServiceClient.ServiceClient {
     constructor(credentials: msRestAzure.ApplicationTokenCredentials, subscriptionId: string, options?: any) {
         super(credentials, subscriptionId);
 
-        this.apiVersion = '2017-05-10';
+        this.apiVersion = '2016-06-01';
```
Comment (sachinma) [inline Tasks/Common/azure-arm-rest/azure-arm-resource.ts]: You shouldn't reduce it. We have updated the version due to a new feature in the new version of the API. I would recommend passing a new isAzureStackEnvironment Boolean in the options object and based on that we can select the API-version

File: Tasks/AzureResourceGroupDeployment/main.ts [hunk 22]
```diff
@@ -6,25 +6,27 @@ import virtualMachine = require("./operations/VirtualMachine");
...
-        case "Start":
-        case "Stop":
-        case "Restart":
-        case "Delete":
-        case "StopWithDeallocate":
-            return virtualMachineOperation.execute();
-        default:
-            throw tl.loc("InvalidAction", taskParameters.action);
-    }
+    return new deployAzureRG.AzureRGTaskParameters().getAzureRGTaskParameters().then((taskParameters) => {
```
Comment (ds-ms) [inline Tasks/AzureResourceGroupDeployment/main.ts]: This chain seems to be unnecessarily too long, you could add a global func called getAzureRGTaskParameters inside that module and call it like, deployAzureRG.getAzureRGTaskParameters()

File: Tasks/AzureResourceGroupDeployment/models/DeployAzureRG.ts [hunk 23]
```diff
@@ -106,14 +76,71 @@ export class AzureRGTaskParameters {
         }
     }
 
-    private getARMCredentials(connectedService: string): msRestAzure.ApplicationTokenCredentials {
+    private async getARMCredentials(connectedService: string): Promise<msRestAzure.ApplicationTokenCredentials> {
```
Comment (ds-ms) [inline Tasks/AzureResourceGroupDeployment/models/DeployAzureRG.ts]: Consider not using async, instead of await, you could try initializeAzureStackData.then

File: Tasks/Common/azure-arm-rest/azure-arm-network.ts [hunk 24]
```diff
@@ -14,7 +14,7 @@ export class NetworkManagementClient extends azureServiceClient.ServiceClient {
     constructor(credentials: msRestAzure.ApplicationTokenCredentials, subscriptionId, baseUri?: any, options?: any) {
         super(credentials, subscriptionId);
 
-        this.apiVersion = '2016-09-01';
+        this.apiVersion = (credentials.isAzureStackEnvironment) ? '2015-06-15' : '2016-09-01' ;
```
Comment (ds-ms) [inline Tasks/Common/azure-arm-rest/azure-arm-network.ts]: Why the different Api-versions? The latest one doesn't support this?

Comment (ds-ms) [inline Tasks/Common/azure-arm-rest/azure-arm-network.ts]: If it doesn't, can we move it to support the latest one some how?

File: Tasks/Common/azurestack-common/azurestackrestutility.ts [hunk 25]
```diff
@@ -0,0 +1,75 @@
...
+            
+            var endpointUrl =  endpoint.url;
+            endpointUrl += (endpointUrl.lastIndexOf("/") == endpointUrl.length-1) ? "":"/";
+            var domain = "";
+            try {
+                var index = endpointUrl.indexOf('.');
+                domain = endpointUrl.substring(index+1);
+                domain = (domain.lastIndexOf("/") == domain.length-1) ? domain.substring(0, domain.length-1): domain;
+            } catch(error) {
+                deferred.reject(tl.loc("SpecifiedAzureRmEndpointIsInvalid", endpointUrl));
```
Comment (ds-ms) [inline Tasks/Common/azurestack-common/azurestackrestutility.ts]: Make sure you've added this loc string to task

File: Tasks/Common/azure-arm-rest/azure-arm-network.ts (hunk 24 above)
Comment (Ajay-MS) [inline Tasks/Common/azure-arm-rest/azure-arm-network.ts]: Azure Stack is onPrem Azure environment. It's API versions are not is sync with Azure API versions.

File: Tasks/Common/azurestack-common/azurestackrestutility.ts (hunk 25 above)
Comment (Ajay-MS) [inline Tasks/Common/azurestack-common/azurestackrestutility.ts]: This string is present in module.json file of azurestack-common module.

Comment (ds-ms) [inline Tasks/Common/azurestack-common/azurestackrestutility.ts]: That works in case of Powershell, but it doesn't in node.

Comment (Ajay-MS) [inline Tasks/Common/azurestack-common/azurestackrestutility.ts]: This seems little odd. Even though string has been added successfully to module resource file. Added string to tasks.json for now.

File: Tasks/AzureResourceGroupDeployment/models/DeployAzureRG.ts (hunk 23 above)
Comment (Ajay-MS) [inline Tasks/AzureResourceGroupDeployment/models/DeployAzureRG.ts]: Closed offline.
